qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -137,99 +160,239 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
+        # Approximation degree is set directly as an attribute on the
+        # instance by the UnitarySynthesis pass here as it's not part of
+        # plugin interface. However if for some reason it's not set assume
+        # it's 1.
+        approximation_degree = getattr(self, "_approximation_degree", 1)
+        basis_gates = options["basis_gates"]
+        coupling_map = options["coupling_map"]
+        natural_direction = options["natural_direction"]
+        pulse_optimize = options["pulse_optimize"]
+        gate_lengths = options["gate_lengths"]
+        gate_errors = options["gate_errors"]
+        qubits = options["qubits"]
+
+        euler_basis = _choose_euler_basis(basis_gates)
+        kak_gate = _choose_kak_gate(basis_gates)
 
-        euler_basis = _choose_euler_basis(self._basis_gates)
-        kak_gate = _choose_kak_gate(self._basis_gates)
         decomposer1q, decomposer2q = None, None
         if euler_basis is not None:
             decomposer1q = one_qubit_decompose.OneQubitEulerDecomposer(euler_basis)
         if kak_gate is not None:
             decomposer2q = TwoQubitBasisDecomposer(
-                kak_gate, euler_basis=euler_basis, pulse_optimize=self._pulse_optimize
+                kak_gate, euler_basis=euler_basis, pulse_optimize=pulse_optimize
             )
 
-        for node in dag.named_nodes(*self._synth_gates):
-            if self._basis_gates and node.name in self._basis_gates:
-                continue
-            synth_dag = None
-            wires = None
-            if len(node.qargs) == 1:
-                if decomposer1q is None:
-                    continue
-                synth_dag = circuit_to_dag(decomposer1q._decompose(node.op.to_matrix()))
-            elif len(node.qargs) == 2:
-                if decomposer2q is None:
-                    continue
-                synth_dag, wires = self._synth_natural_direction(node, dag, decomposer2q)
-            else:
-                synth_dag = circuit_to_dag(isometry.Isometry(node.op.to_matrix(), 0, 0).definition)
-
-            dag.substitute_node_with_dag(node, synth_dag, wires=wires)
+        synth_dag = None
+        wires = None
+        if unitary.shape == (2, 2):
+            if decomposer1q is None:
+                return None
+            synth_dag = circuit_to_dag(decomposer1q._decompose(unitary))
+        elif unitary.shape == (4, 4):
+            if decomposer2q is None:
+                return None
+            synth_dag, wires = self._synth_natural_direction(
+                unitary,
+                coupling_map,
+                qubits,
+                decomposer2q,
+                gate_lengths,
+                gate_errors,
+                natural_direction,
+                approximation_degree,
+                pulse_optimize,
+            )
+        else:
+            synth_dag = circuit_to_dag(isometry.Isometry(unitary, 0, 0).definition)
 
-        return dag
+        return synth_dag, wires
```

Comment:
This uses slightly undocumented behaviour in the plugin interface, since it always returns a tuple, except sometimes `wires` is None.  If you just read the docstring of the plugin `run` method, you might expect that the return should be `return (synth_dag, wires) if wires is not None else synth_dag` - it's not explicitly mentioned that `wires=None` is sufficient as well (though the continuation function _is_ linked, so I suppose you can know it).

Reply:
I'll add this to the docstring for the run method to make it clear that wires can be None too

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/6124/commits/5d8539ac37300473270b470574aa8879c975e2f9